beacon-chain/db/filesystem/pruner.go

```diff
@@ -0,0 +1,250 @@
+package filesystem
+
+import (
+	"encoding/binary"
+	"io"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pkg/errors"
+	fieldparams "github.com/prysmaticlabs/prysm/v4/config/fieldparams"
+	"github.com/prysmaticlabs/prysm/v4/config/params"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+	"github.com/prysmaticlabs/prysm/v4/time/slots"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/afero"
+)
+
+const retentionBuffer primitives.Epoch = 2
+
+var (
+	errPruningFailures = errors.New("blobs could not be pruned for some roots")
+)
+
+type blobPruner struct {
+	sync.Mutex
+	prunedBefore atomic.Uint64
+	windowSize   primitives.Slot
+	slotMap      *slotForRoot
+	fs           afero.Fs
+}
+
+func newblobPruner(fs afero.Fs, retain primitives.Epoch) (*blobPruner, error) {
+	r, err := slots.EpochStart(retain + retentionBuffer)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not set retentionSlots")
+	}
+	return &blobPruner{fs: fs, windowSize: r, slotMap: newSlotForRoot()}, nil
+}
+
+// notify updates the pruner's view of root->blob mappings. This allows the pruner to build a cache
+// of root->slot mappings and decide when to evict old blobs based on the age of present blobs.
+func (p *blobPruner) notify(root [32]byte, latest primitives.Slot) {
+	p.slotMap.ensure(rootString(root), latest)
+	pruned := uint64(windowMin(latest, p.windowSize))
+	if p.prunedBefore.Swap(pruned) == pruned {
+		return
+	}
+	go func() {
+		if err := p.prune(primitives.Slot(pruned)); err != nil {
+			log.WithError(err).Errorf("failed to prune blobs from slot %d", latest)
+		}
+	}()
+}
+
+func windowMin(latest primitives.Slot, offset primitives.Slot) primitives.Slot {
+	// Safely compute the first slot in the epoch for the latest slot
+	latest = latest - latest%params.BeaconConfig().SlotsPerEpoch
+	if latest < offset {
+		return 0
+	}
+	return latest - offset
+}
+
+// Prune prunes blobs in the base directory based on the retention epoch.
+// It deletes blobs older than currentEpoch - (retentionEpochs+bufferEpochs).
+// This is so that we keep a slight buffer and blobs are deleted after n+2 epochs.
+func (p *blobPruner) prune(pruneBefore primitives.Slot) error {
+	p.Lock()
+	defer p.Unlock()
+	log.Debug("Pruning old blobs")
```

Comment:
Is a pruning log necessary? It guess we want to signal something has started?
```
[2024-01-07 08:10:47]  INFO initial-sync: Processing block batch of size 54 starting from  0x1d03ca02... 278688/281753 - estimated time remaining 18m55s blocksPerSecond=2.7 peers=32
[2024-01-07 08:10:47] DEBUG Pruning old blobs
[2024-01-07 08:10:47] DEBUG Pruned old blobs blobsPruned=10 duration=23.337542ms prunedUpToEpoch=4611
[2024-01-07 08:10:48] DEBUG Pruning old blobs
[2024-01-07 08:10:48] DEBUG Pruned old blobs blobsPruned=27 duration=25.083167ms prunedUpToEpoch=4612
[2024-01-07 08:10:49]  INFO initial-sync: Processing block batch of size 57 starting from  0x5e91020f... 278752/281754 - estimated time remaining 9m0s blocksPerSecond=5.5 peers=32
[2024-01-07 08:10:49] DEBUG Pruning old blobs
[2024-01-07 08:10:49] DEBUG Pruned old blobs blobsPruned=22 duration=25.066792ms prunedUpToEpoch=4613
[2024-01-07 08:10:49] DEBUG Pruning old blobs
[2024-01-07 08:10:49] DEBUG Pruned old blobs blobsPruned=34 duration=25.282125ms prunedUpToEpoch=4614
[2024-01-07 08:10:50]  INFO initial-sync: Processing block batch of size 58 starting from  0xb88bd110... 278816/281754 - estimated time remaining 5m47s blocksPerSecond=8.4 peers=32
[2024-01-07 08:10:50] DEBUG Pruning old blobs
[2024-01-07 08:10:50] DEBUG Pruned old blobs blobsPruned=35 duration=33.01875ms prunedUpToEpoch=4615
[2024-01-07 08:10:51] DEBUG Pruning old blobs
[2024-01-07 08:10:51] DEBUG Pruned old blobs blobsPruned=31 duration=23.6385ms prunedUpToEpoch=4616
[2024-01-07 08:10:52]  INFO initial-sync: Processing block batch of size 61 starting from  0x4c4938b8... 278880/281754 - estimated time remaining 4m9s blocksPerSecond=11.5 peers=32
[2024-01-07 08:10:52] DEBUG Pruning old blobs
[2024-01-07 08:10:52] DEBUG Pruned old blobs blobsPruned=35 duration=24.930666ms prunedUpToEpoch=4617
[2024-01-07 08:10:52] DEBUG Pruning old blobs
[2024-01-07 08:10:53] DEBUG Pruned old blobs blobsPruned=21 duration=23.224584ms prunedUpToEpoch=4618
[2024-01-07 08:10:53]  INFO initial-sync: Processing block batch of size 57 starting from  0xeda74edc... 278944/281754 - estimated time remaining 3m15s blocksPerSecond=14.3 peers=32
[2024-01-07 08:10:54] DEBUG Pruning old blobs
[2024-01-07 08:10:54] DEBUG Pruned old blobs blobsPruned=25 duration=25.345042ms prunedUpToEpoch=4619
[2024-01-07 08:10:54] DEBUG Pruning old blobs
[2024-01-07 08:10:54] DEBUG Pruned old blobs blobsPruned=5 duration=22.283542ms prunedUpToEpoch=4620
[2024-01-07 08:10:55]  INFO initial-sync: Processing block batch of size 60 starting from  0xee020db7... 279009/281754 - estimated time remaining 2m38s blocksPerSecond=17.4 peers=32
```

Reply:
I think we had this before, I'm ambivalent about keeping it.

Reply:
I thought I removed it here https://github.com/prysmaticlabs/prysm/pull/13417

Reply:
Ah I copied the code from an earlier commit. I'll axe it here.